libvast/builtins/operators/summarize.cpp

```diff
@@ -321,212 +209,368 @@ struct group_by_key_equal {
+    reusable_key_view.resize(bound.group_by_columns.size(), {});
+    // Returns the group that the given row belongs to, creating new groups
+    // whenever necessary.
+    auto find_or_create_bucket = [&](int64_t row) -> caf::expected<bucket*> {
+      for (size_t col = 0; col < bound.group_by_columns.size(); ++col) {
+        if (bound.group_by_columns[col]) {
+          VAST_ASSERT(group_by_arrays[col].has_value());
+          reusable_key_view[col] = value_at(bound.group_by_columns[col]->type,
+                                            **group_by_arrays[col], row);
+        } else {
+          VAST_ASSERT(!group_by_arrays[col].has_value());
+          reusable_key_view[col] = caf::none;
+        }
+      }
+      if (auto it = buckets.find(reusable_key_view); it != buckets.end()) {
+        auto&& bucket = *it->second;
+        // Check that the group-by values also have matching types.
+        VAST_ASSERT(bucket.group_by_types.size()
+                    == bound.group_by_columns.size());
+        for (auto&& [existing, other] :
+             detail::zip{bucket.group_by_types, bound.group_by_columns}) {
+          if (!other) {
+            // If this group-by column does not exist in the input schema, we
+            // assume `null` values and ignore its type.
+            continue;
+          }
+          if (other->type != existing) {
+            // If the types mismatch, we have found equal values with not-equal
+            // types. This can happen e.g. for `null` values. TODO: Prune type?
+            if (existing == type{}) {
+              // In this case, we found a bucket for a schema which does not
+              // have the group-by extractor at all (hence the type is also
+              // `null`). Thus, we can just change the type.
+              static_assert(std::is_reference_v<decltype(existing)>);
+              existing = other->type;
+            } else {
+              return caf::make_error(
+                ec::type_clash,
+                fmt::format("summarize found matching group for key `{}`, but "
+                            "the existing type `{}` clashes with `{}`",
+                            reusable_key_view, existing, other->type));
+            }
+          }
+        }
+        // Check that the aggregation extractors have the same type.
+        VAST_ASSERT(bucket.functions.size()
+                    == bound.aggregation_columns.size());
+        for (auto&& [func, column] :
+             detail::zip{bucket.functions, bound.aggregation_columns}) {
+          auto func_type = func ? func->input : type{};
+          auto col_type = column ? column->type : type{};
+          if (func_type != col_type) {
+            return caf::make_error(
+              ec::type_clash,
+              fmt::format("summarize aggregation function for group `{}` "
+                          "expected type `{}`, but got `{}`",
+                          reusable_key_view, func_type, col_type));
```

Comment:
As discussed in Slack, this has to change for schemas where the field does not exist.

Reply:
The behavior for this and a few other things changed dramatically since your last review. Basically, `summarize` will never abort the pipeline execution anymore (except under catastrophic circumstances). Please take a look again. 

I think this behavior here is what we want, it works out quite nicely. Feel free to resolve the remaining two conversations if you are also happy with it (I enabled auto-merge).